rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala

```diff
@@ -257,6 +257,76 @@ object ProcNormalizeMatcher {
+        // To handle the most common case where we can sort the binds because
+        // they're from different sources, Each channel read starts its free variables at 0.
+        // We check for overlap at the end before sorting.
+
+        // We split this into parts. First we process all the sources, then we process all the bindings.
+        def processSources(bindings: List[(List[Name], Name)])
+          : (List[(List[Name], Channel)], DebruijnLevelMap[VarSort]) = {
+          val initAcc = (List[(List[Name], Channel)](), input.knownFree)
+          val foldResult = (initAcc /: bindings)((acc, e) => {
+            val sourceResult =
+              NameNormalizeMatcher.normalizeMatch(e._2, NameVisitInputs(input.env, acc._2))
+            ((e._1, sourceResult.chan) :: acc._1, sourceResult.knownFree)
+          })
+          (foldResult._1.reverse, foldResult._2)
+        }
+        def processBindings(bindings: List[(List[Name], Channel)])
+          : List[(List[Channel], Channel, DebruijnLevelMap[VarSort])] =
+          bindings map {
+            case (names: List[Name], chan: Channel) => {
+              val initAcc = (List[Channel](), DebruijnLevelMap[VarSort]())
+              val formalsResults = (initAcc /: names)(
+                (acc, n: Name) => {
+                  val result =
+                    NameNormalizeMatcher.normalizeMatch(n,
+                                                        NameVisitInputs(DebruijnLevelMap(), acc._2))
+                  (result.chan :: acc._1, result.knownFree)
+                }
+              )
+              (formalsResults._1.reverse, chan, formalsResults._2)
+            }
+          }
+        val (bindings, persistent) = p.receipt_ match {
+          case rl: ReceiptLinear =>
+            rl.receiptlinearimpl_ match {
+              case ls: LinearSimple =>
+                (ls.listlinearbind_.asScala.toList.map {
+                  case lbi: LinearBindImpl => (lbi.listname_.asScala.toList, lbi.name_)
+                  case _                   => throw new Error("Unexpected LinearBind production.")
+                }, false)
+              case _ => throw new Error("Unexpected LinearReceipt production.")
+            }
+          case rl: ReceiptRepeated =>
+            rl.receiptrepeatedimpl_ match {
+              case ls: RepeatedSimple =>
+                (ls.listrepeatedbind_.asScala.toList.map {
+                  case lbi: RepeatedBindImpl => (lbi.listname_.asScala.toList, lbi.name_)
+                  case _                     => throw new Error("Unexpected RepeatedBind production.")
+                }, true)
+              case _ => throw new Error("Unexpected RepeatedReceipt production.")
+            }
+        }
+        val (sources, thisLevelFree) = processSources(bindings)
+        val receipts                 = ReceiveSortMatcher.preSortBinds(processBindings(sources))
+        val groupedFrees = (DebruijnLevelMap[VarSort]() /: receipts)((env, receipt) =>
```

Comment:
Another nitpick/preference but I would call this "renumberedFrees".

Reply:
Oh I just realized that its named "groupedFrees" because these frees got are now "grouped" together and renumbered. It now is just as appealing of a name to me: take your pick.

Reply:
Hmm what about "combinedFrees". I think initially "grouped" made me think of something that is divided into many things.